ltiauthenticator/utils.py

```diff
@@ -0,0 +1,99 @@
+import re
+
+from tornado.log import app_log
+from tornado.web import RequestHandler
+
+
+def normalize_string(name: str) -> str:
+    """
+    Function used to strip special characters and convert strings
+    to docker container and Kubernetes compatible names.
+
+    Args:
+        name: The string to normalize
+
+    Returns:
+        normalized_name: The normalized string
+    """
+    if not name:
+        raise ValueError("Name is empty")
+    # truncate name after 30th character
+    name = (name[:25] + "") if len(name) > 30 else name
+    # remove special characters
+    name = re.sub(r"[^\w-]+", "", name)
+    # if the first character is any of _.- remove it
+    name = name.lstrip("_.-")
+    # convert to lower case
+    name = name.lower()
+    # limit course_id to 25 characters, since its used for o/s username
+    # in jupyter/docker-stacks compatible grader notebook (NB_USER)
+    normalized_name = name[0:25]
+    app_log.debug("String normalized to %s" % normalized_name)
+    return normalized_name
+
+
+def email_to_username(email: str) -> str:
```

Comment:
Can we use escapism for this too?

Will there be username clashes if we have two users from different domains but same username?

Reply:
@yuvipanda escapism would be a good option here too, I believe. To answer your second question, yes there would be clashes. We have gotten around this by concatenating the username with the user's unique identifier from the LMS, for example, foobar-123 and foobar-456 would have the same username extrapolated from the email (for example) but then the username is differentiated by also including the user_id value.

Generally speaking, this is something we add to the authenticator logic before creating the Authenticator's return dict since this utility function _could_ be used for just normalizing an email (although unlikely).